### PR TITLE
Feature eos_staticroute

### PR DESCRIPTION
**Pull request by jerearista**

### CODE REVIEW

File: lib/puppet/provider/eos_staticroute/default.rb [hunk 1]
```diff
@@ -30,38 +30,44 @@
...
-
-    result.split(/\n/).map do |entry|
-      parts = entry.split
-      provider_hash = { name: parts[2], ensure: :present }
-      provider_hash[:next_hop] = parts[3]
-      provider_hash[:route_name] = ''
-      new(provider_hash)
+    routes = node.api('staticroutes').getall
+    return [] if routes.empty?
+    routes.each_with_object([]) do |attrs, arry|
```
Comment (devrobo) [inline lib/puppet/provider/eos_staticroute/default.rb]: If routes is empty doesn't this instantiate arry with []? (i.e. do we need line 50)

File: lib/puppet/provider/eos_staticroute/default.rb [hunk 2]
```diff
@@ -30,38 +30,44 @@
...
 
-  def self.prefetch(resources)
-    provider_hash = instances.each_with_object({}) do |provider, hsh|
-      hsh[provider.name] = provider
-    end
-
-    resources.each_pair do |name, resource|
-      resource.provider = provider_hash[name] if provider_hash[name]
-    end
+    #   pry(main)> node.api('staticroutes').getall
```
Comment (devrobo) [inline lib/puppet/provider/eos_staticroute/default.rb]: Did you want to commit this commented code?

File: lib/puppet/provider/eos_staticroute/default.rb [hunk 3]
```diff
@@ -30,38 +30,44 @@
...
+    #     :tag=>"3",
+    #     :name=>"fred"},
+    #    {:destination=>"192.0.3.0/24",
+    #     :nexthop=>"192.0.3.1",
+    #     :distance=>"1",
+    #     :tag=>"0",
+    #     :name=>"dummy2"}]
   end
 
   def initialize(resource = {})
```
Comment (devrobo) [inline lib/puppet/provider/eos_staticroute/default.rb]: I think the initialize needs to look like this:

  def initialize(resource = {})
    super(resource)
    @property_flush = {}
  end

File: lib/puppet/provider/eos_staticroute/default.rb [hunk 4]
```diff
@@ -30,38 +30,44 @@
...
-    result.split(/\n/).map do |entry|
-      parts = entry.split
-      provider_hash = { name: parts[2], ensure: :present }
-      provider_hash[:next_hop] = parts[3]
-      provider_hash[:route_name] = ''
-      new(provider_hash)
+    routes = node.api('staticroutes').getall
+    return [] if routes.empty?
+    routes.each_with_object([]) do |attrs, arry|
+      name = namevar(attrs[:destination], attrs[:nexthop])
```
Comment (devrobo) [inline lib/puppet/provider/eos_staticroute/default.rb]: Is attrs[:destination] and attrs[:nexthop] always defined?

File: lib/puppet/type/eos_staticroute.rb [hunk 5]
```diff
@@ -62,4 +89,32 @@
       end
     end
   end
+
+  newproperty(:distance) do
+    @doc = <<-EOS
+      Administrative distance of the route
+    EOS
+
+    newvalues(1..255)
```
Comment (devrobo) [inline lib/puppet/type/eos_staticroute.rb]: Why is this here?

File: lib/puppet/type/eos_staticroute.rb [hunk 6]
```diff
@@ -62,4 +89,32 @@
...
+      end
+    end
+  end
+
+  newproperty(:tag) do
+    @doc = <<-EOS
+      Route tag
+    EOS
+
+    newvalues(1..255)
```
Comment (devrobo) [inline lib/puppet/type/eos_staticroute.rb]: Why is this here?

File: spec/unit/puppet/type/eos_staticroute_spec.rb [hunk 7]
```diff
@@ -50,6 +53,16 @@
...
+    subject { described_class.attrclass(attribute) }
+
+    include_examples 'property'
+    include_examples '#doc Documentation'
+    include_examples 'accepts values without munging', %w(1 255)
+    include_examples 'rejects values', [[0], { two: :three }]
+  end
+
   describe 'route_name' do
     let(:attribute) { :route_name }
```
Comment (devrobo) [inline spec/unit/puppet/type/eos_staticroute_spec.rb]: Lines 67-73 can be replaced with:
include_examples 'string', attribute: :route_name

File: spec/unit/puppet/type/eos_staticroute_spec.rb [hunk 8]
```diff
@@ -50,6 +53,16 @@
     include_examples 'rejects values', [[1], { two: :three }]
   end
 
+  describe 'distance' do
+    let(:attribute) { :distance }
+    subject { described_class.attrclass(attribute) }
+
+    include_examples 'property'
+    include_examples '#doc Documentation'
+    include_examples 'accepts values without munging', %w(1 255)
```
Comment (devrobo) [inline spec/unit/puppet/type/eos_staticroute_spec.rb]: Can 62 and 63 be replaced with:

    include_examples 'numeric parameter', min: 1, max: 255
    include_examples 'rejects values', [0, 256]

Let me know the outcome. Will the example pass in a string or an int?

File: lib/puppet/provider/eos_staticroute/default.rb (hunk 1 above)
Comment (jerearista) [inline lib/puppet/provider/eos_staticroute/default.rb]: Agreed.  removed line

File: lib/puppet/provider/eos_staticroute/default.rb (hunk 2 above)
Comment (jerearista) [inline lib/puppet/provider/eos_staticroute/default.rb]: removed

File: lib/puppet/provider/eos_staticroute/default.rb (hunk 4 above)
Comment (jerearista) [inline lib/puppet/provider/eos_staticroute/default.rb]: Yes.   EOS CLI won't take an entry without those required fields.

File: lib/puppet/type/eos_staticroute.rb (hunk 5 above)
Comment (jerearista) [inline lib/puppet/type/eos_staticroute.rb]: standard route metric definition.
